maps-compose/src/main/java/com/google/maps/android/compose/GoogleMap.kt

```diff
@@ -157,72 +299,57 @@ internal suspend inline fun disposingComposition(factory: () -> Composition) {
- */
-@Composable
-private fun MapLifecycle(mapView: MapView) {
-    val context = LocalContext.current
-    val lifecycle = LocalLifecycleOwner.current.lifecycle
-    val previousState = remember { mutableStateOf(Lifecycle.Event.ON_CREATE) }
-    DisposableEffect(context, lifecycle, mapView) {
-        val mapLifecycleObserver = mapView.lifecycleObserver(previousState)
-        val callbacks = mapView.componentCallbacks()
-
-        lifecycle.addObserver(mapLifecycleObserver)
-        context.registerComponentCallbacks(callbacks)
+private class MapViewLifecycleController(
+    isMapReused: Boolean,
+    private val mapView: MapView
+) {
+    private var previousState = if (isMapReused)
+        Lifecycle.Event.ON_STOP else
+        Lifecycle.Event.ON_CREATE
 
-        onDispose {
-            lifecycle.removeObserver(mapLifecycleObserver)
-            context.unregisterComponentCallbacks(callbacks)
+    var lifecycle: Lifecycle? = null
+        set(value) {
+            if (field !== value) {
+                field?.removeObserver(observer)
+                value?.addObserver(observer)
+                field = value
+            }
         }
+
+    fun onDestroy() {
+        lifecycle = null
+        mapView.onDestroy()
     }
-    DisposableEffect(mapView) {
-        onDispose {
-            mapView.onDestroy()
-            mapView.removeAllViews()
-        }
+
+    fun onLifecycleDetached() {
+        lifecycle = null
+        mapView.onStop()
     }
-}
 
-private fun MapView.lifecycleObserver(previousState: MutableState<Lifecycle.Event>): LifecycleEventObserver =
-    LifecycleEventObserver { _, event ->
-        event.targetState
+    private val observer = LifecycleEventObserver { _, event ->
         when (event) {
             Lifecycle.Event.ON_CREATE -> {
                 // Skip calling mapView.onCreate if the lifecycle did not go through onDestroy - in
                 // this case the GoogleMap composable also doesn't leave the composition. So,
                 // recreating the map does not restore state properly which must be avoided.
-                if (previousState.value != Lifecycle.Event.ON_STOP) {
-                    this.onCreate(Bundle())
+                if (previousState != Lifecycle.Event.ON_STOP) {
+                    mapView.onCreate(Bundle())
                 }
             }
 
-            Lifecycle.Event.ON_START -> this.onStart()
-            Lifecycle.Event.ON_RESUME -> this.onResume()
-            Lifecycle.Event.ON_PAUSE -> this.onPause()
-            Lifecycle.Event.ON_STOP -> this.onStop()
+            Lifecycle.Event.ON_START -> mapView.onStart()
+            Lifecycle.Event.ON_RESUME -> mapView.onResume()
+            Lifecycle.Event.ON_PAUSE -> mapView.onPause()
+            Lifecycle.Event.ON_STOP -> mapView.onStop()
             Lifecycle.Event.ON_DESTROY -> {
-                //handled in onDispose
+                // Handled in AndroidView onRelease
```

Comment:
Not entirely sure what you mean.
Do you mean that you think that onRelease will not be invoked when parent lifecycle reaches onDestroy()?

Reply:
My general thinking is that androidx lifecycle and AndroidView lifecycle are independent, in principle.

For example, I'd think that an Activity reaching DESTROYED state would not cause AndroidView `onRelease` to be called, because the View is not actually removed from the composition. However, it would cause Lifecycle.Event.ON_DESTROY to be called.

Reply:
onDestroy is now invoked by the LifecycleEventObserver. This observer is only unregistered in AndroidView onRelease now.
onDestroy is also invoked in onRelease (if it's not already destroyed).